airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,289 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Callable, Iterable, Mapping, Optional, Tuple
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.stream_slicers.stream_slicer import StreamSlicer
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState
+
+
+class PerPartitionKeySerializer:
+    """
+    We are concerned of the performance of looping through the `states` list and evaluating equality on the partition. To reduce this
+    concern, we wanted to use dictionaries to map `partition -> cursor`. However, partitions are dict and dict can't be used as dict keys
+    since they are not hashable. By creating hashable key and values, we can implement __hash__ for dict and then use them as dict keys.
+    Note that those dicts shouldn't be modified has their hash would change. For our case, it should be fine as partitions can be seen as
+    immutable.
+    """
+
+    @staticmethod
+    def to_partition_key(to_serialize: Any) -> Tuple:
+        return tuple(PerPartitionKeySerializer._to_partition_key(to_serialize))
+
+    @staticmethod
+    def _to_partition_key(to_serialize: Any, previous_keys=()):
+        if isinstance(to_serialize, dict):
+            for key in sorted(to_serialize.keys()):
+                yield from PerPartitionKeySerializer._to_partition_key(to_serialize[key], previous_keys=(*previous_keys, key))
+        elif isinstance(to_serialize, list):
+            if isinstance(to_serialize[0], dict):
+                yield *previous_keys, tuple([PerPartitionKeySerializer.to_partition_key(dictionary) for dictionary in to_serialize])
+            else:
+                yield *previous_keys, tuple(to_serialize)
+        else:
+            yield *previous_keys, to_serialize
+
+    @staticmethod
+    def to_partition(to_deserialize: Any):
+        root = {}
+        for flattened_entry in to_deserialize:
```

Comment:
I suppose I wouldn't call this `flattened` since the tuples are nested. (I was thinking of a flattened dictionary as `{'a': {'a1': 1, 'a2': 2}, 'b': 2}` => `((("a", "a1"), 1), (("a", "a2"), 2), (("b",), 2))`, but don't see a need to do this as your nested tuple solution seems fine and more closely resembles the dictionary.) How about just `entry`?

Reply:
Okay cool, in that case it makes sense to me to keep `flattened_entry` at the root.